plugins/modules/artisan.py

```diff
@@ -0,0 +1,324 @@
+
+
+def has_changed(message):
+    # message fragments that Laravel gives us
+    artisan_messages = [
+        'No publishable',  # resources for tag
+        'Nothing to',  # migrate,...
+        'already',  # up, down, exists,...
+        'Not modified',
+        'No changes'
+    ]
+
+    for info in artisan_messages:
+        if info in message:
+            return False
+
+    return True
+
+
+def get_help(module, command):
+    # get all available options and arguments from an artisan command using artisan help to json
+    rc, out, err = artisan_command(module, 'help %s' % command, options='--format=json --no-ansi')
+
+    # no command, no help, no json
+    if rc != 0:
+        module.fail_json(msg='%s is not part of the command definitions.' % command)
+
+    return module.from_json(out)
+
+
+def get_options(module, command_help):
+    # set available command options
+    available_options = command_help['definition']['options']
+
+    # merge default options
+    command_options = module.params['options'] + [
+        'no-interaction',
+        'no-ansi',
+        'force',
+    ]
+
+    options = []
+
+    # check availability and format options
+    for option in command_options:
+        option = list(map(str.strip, option.lstrip('-').split('=')))
+        if option[0] in available_options:
+            options.append('--%s' % '='.join(option))
+
+    # return the unique options turnkey ready as a string
+    return ' '.join(list(set(options))).strip()
+
+
+def get_arguments(module, command_help):
+    # set available command arguments
+    available_arguments = command_help['definition']['arguments']
+
+    # mostly we have no arguments
+    arguments = ''
+
+    # if arguments expected for command, join from params
+    if len(available_arguments) > 0:
+        arguments = ' '.join(module.params['args'])
+
+    # return arguments turnkey ready as a string
+    return arguments.strip()
+
+
+def get_artisan(module):
+    artisan = '%s/artisan' % os.path.abspath(module.params['working_dir'])
+
+    # check artisan executable path
+    if not os.path.isfile(artisan):
+        module.fail_json(msg='The artisan executable is not present.', artisan=artisan)
+
+    return artisan
+
+
+def get_php(module):
+    # in most cases, PHP is available by default
+    if module.params['php_path'] is None:
+        php = module.get_bin_path('php', True, ['/usr/local/bin'])
+    else:
+        php = os.path.abspath(module.params['php_path'])
+
+    # check php executable path
+    if not os.path.isfile(php):
+        module.fail_json(msg='The php executable is not present.', php=php)
+
+    return php
+
+
+def artisan_command(module, command, options, arguments=''):
+    # this is the artisan executable in the project root
+    artisan = get_artisan(module)
+
+    # PHP is essential for successful execution
+    php = get_php(module)
+
+    # combine the entire command
+    command = '%s %s %s %s %s' % (php, artisan, command, options, arguments)
```

Comment:
Please do not use this form, preferring instead the form in which `command` ends up being a `list` (or `tuple` for that matter). By doing that we decrease the risk of shell script injection - this string way actually uses the system shell to execute the command - you can imagine the fun if one of the params is `; rm /etc/passwd; true`.

This is particularly crucial in this module where we have free-form strings coming in the `args` parameter and added to the command line.

Reply:
The arguments are previously compared with those available and filtered. But the Command itself would be vulnerable. I'll change that.